### PR TITLE
feat: add Kitty Keyboard Protocol support

### LINKED ISSUE #1217: Kitty Keyboard Protocol support
Perhaps https://sw.kovidgoyal.net/kitty/keyboard-protocol/

### PR DESCRIPTION
**Pull request by gnodet**
## Summary

Implements the [Kitty Keyboard Protocol](https://sw.kovidgoyal.net/kitty/keyboard-protocol/) (fixes #1217), a progressive enhancement that replaces legacy ambiguous terminal keyboard encodings with unambiguous `CSI … u` sequences.

- **Terminal detection & lifecycle**: `hasKittyKeyboardSupport()` probes the terminal via `CSI ? u` + DA1 sentinel; `setKittyKeyboardMode(flags)` / `resetKittyKeyboardMode()` push/pop enhancement flags
- **Full CSI … u parsing**: `KeyParser` handles all protocol fields — key codes, modifiers (including Super/Hyper/Meta/CapsLock/NumLock), event types (press/repeat/release), alternate keys (shifted/base layout), and associated text
- **Extended KeyEvent model**: new `EventType` enum, 5 additional `Modifier` values, `keyCode`, `shiftedKeyCode`, `baseLayoutKeyCode`, and `associatedText` fields — all backward-compatible
- **LineReader integration**: `Option.KITTY_KEYBOARD` pushes flag 1 (disambiguate) during `readLine()` and pops on exit; kitty-variant bindings registered for Ctrl+letter, Alt+letter, Shift+Enter, Ctrl+Backspace, etc.
- **47 new tests** covering sequence parsing, modifier encoding, event types, and protocol helpers

Only flag 1 (disambiguate) is enabled by default — the parser supports all flags (1–16) so applications can push higher flags directly.

Supported by kitty, Alacritty, foot, Ghostty, iTerm2, WezTerm, Windows Terminal, Rio, Warp, and xterm.js.

## Test plan

- [x] `KittyKeyboardSupportTest`: protocol constants, sequence builders, flag response parsing (11 tests)
- [x] `KeyParserKittyTest`: CSI u parsing — characters, specials, modifiers, event types, alternate keys, associated text, function keys, edge cases (36 tests)
- [x] Full `terminal` module test suite passes (394 tests, 0 failures)
- [x] Full `reader` module test suite passes (209 tests, 0 failures)
- [x] Full project builds successfully
- [ ] Manual testing with a kitty-protocol-capable terminal (kitty/Ghostty/WezTerm)

## Summary by CodeRabbit

* **New Features**
  * Added Kitty Keyboard Protocol support for enhanced key reporting, including event types, modifiers, alternate key codes, and associated text.
  * Added terminal and line-reader controls for enabling Kitty keyboard modes when supported.
  * Added terminal resize reporting, automatic dimension updates, and resize event notifications.
  * Expanded keymaps and input parsing for Kitty-encoded keys, keypad, media, lock, and modifier keys.
  * Added an interactive key-event viewer demonstration.
* **Bug Fixes**
  * Improved terminal cleanup by reliably resetting keyboard modes and resize tracking.
  * Safer fallback highlighting for control characters and reduced ambiguous-key response delay.

### CODE REVIEW

File: terminal/src/main/java/org/jline/terminal/KeyParser.java [hunk 1]
```diff
@@ -334,4 +339,244 @@ private static KeyEvent parseControlCharacter(char ch, String sequence) {
...
+                    rawSequence,
+                    eventType,
+                    keyCode,
+                    shiftedKeyCode,
+                    baseLayoutKeyCode,
+                    associatedText);
+        }
+
+        // Printable characters (Unicode codepoints below Private Use Area)
+        if (keyCode >= 32 && keyCode < 57344) {
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/KeyParser.java]: Good catch. Fixed — changed the condition to exclude only the BMP Private Use Area (U+E000–U+F8FF) where the kitty protocol defines its functional key codes. Non-BMP Unicode characters (emoji, CJK Extension B, etc.) are now correctly classified as `Character`. Added tests for both cases.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 2]
```diff
@@ -6947,6 +6959,71 @@ private void bindArrowKeys(KeyMap<Binding> map) {
...
+            String legacySeq = Character.toString((char) ctrlCode);
+            Object bound = map.getBound(legacySeq);
+            if (bound instanceof Binding) {
+                map.bind((Binding) bound, KittyKeyboardSupport.ctrlKey(c));
+            }
+        }
+
+        // With kitty flag 1, Escape sends CSI 27u instead of byte 0x1B
+        bind(map, VI_CMD_MODE, "\033[27u");
+
```
[the remark this answers is not in the document]
Comment (gnodet) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java]: Valid point. Moved the kitty Escape (`CSI 27u`) → `VI_CMD_MODE` binding out of `bindKittyKeys()` (which runs for all keymaps) and applied it only to `VIINS` and `VICMD`. The emacs keymap no longer gets this binding.

File: terminal/src/main/java/org/jline/terminal/impl/AbstractTerminal.java [hunk 3]
```diff
@@ -780,6 +791,89 @@ public boolean setGraphemeClusterMode(boolean enable, boolean force) {
...
+        return false;
+    }
+
+    @Override
+    public boolean resetKittyKeyboardMode() {
+        if (kittyKeyboardActive) {
+            writer().write(KittyKeyboardSupport.popFlags());
+            writer().flush();
+            kittyKeyboardActive = false;
+            return true;
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/impl/AbstractTerminal.java:1051]: Good catch. `setKittyKeyboardMode()` now pops the previous flags before pushing new ones when already active, keeping the terminal's stack balanced. A second call with different flags produces pop+push (stack stays at depth 1), and `resetKittyKeyboardMode()` cleanly pops the single entry.

File: terminal/src/main/java/org/jline/terminal/KeyParser.java [hunk 4]
```diff
@@ -334,4 +339,246 @@ private static KeyEvent parseControlCharacter(char ch, String sequence) {
...
+     */
+    private static KeyEvent.Special mapKittySpecialKey(int keyCode) {
+        switch (keyCode) {
+            case 13:
+                return KeyEvent.Special.Enter;
+            case 9:
+                return KeyEvent.Special.Tab;
+            case 27:
+                return KeyEvent.Special.Escape;
+            case 127:
```
Comment (gnodet) [inline terminal/src/main/java/org/jline/terminal/KeyParser.java:857]: Nit: The mappings for keyCode 2 (Insert), 3 (Delete), 5 (PageUp), and 6 (PageDown) in `mapKittySpecialKey()` appear to be unreachable. Per the Kitty Keyboard Protocol specification, these keys are always sent as CSI number `~` sequences (e.g., `CSI 2 ~` for Insert), never as `CSI number u`. Since `parseKittySequence()` only processes sequences ending in `u`, these four cases can never be reached.

The code is not harmful but is slightly misleading — it suggests these keys arrive as CSI `u`, which is incorrect per the spec. Consider removing these four cases and adding a comment explaining that Insert/Delete/PageUp/PageDown use legacy CSI `~` encoding even in kitty mode.

**Review by gnodet — COMMENTED** [review]
Well-structured implementation of the Kitty Keyboard Protocol with comprehensive tests. The core parsing, key binding, and terminal lifecycle management are correct. One minor observation about unreachable code noted below.

The overall design is clean and well-integrated with JLine's existing patterns (Mode enum for probing, push/pop lifecycle, unconditional cleanup in doClose). The decision to only enable flag 1 (Disambiguate) by default is pragmatic. The bindKittyKeys approach of mirroring legacy bindings with kitty equivalents is the right strategy.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: terminal/src/main/java/org/jline/terminal/KeyParser.java [hunk 5]
```diff
@@ -334,4 +339,246 @@ private static KeyEvent parseControlCharacter(char ch, String sequence) {
...
+            case 3:
+                return KeyEvent.Special.Delete;
+            case 5:
+                return KeyEvent.Special.PageUp;
+            case 6:
+                return KeyEvent.Special.PageDown;
+            default:
+                return null;
+        }
+    }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/KeyParser.java:902]: Good catch — the cases for 2/3/5/6 were indeed wrong: those are CSI ~ parameter numbers, not CSI u key codes. Per the spec, Insert/Delete/PageUp/PageDown normally use legacy CSI number ~ encoding even with the kitty protocol active, so those cases were unreachable in the CSI u parser.

Fixed by replacing them with the PUA codepoints the protocol defines (57348–57355), and added Home (57350) and End (57351) while at it. These serve as a defensive fallback in case a terminal sends these keys as CSI u instead of CSI ~. Also corrected the KEY_INSERT/KEY_DELETE/KEY_PAGE_UP/KEY_PAGE_DOWN constants in KittyKeyboardSupport to use the correct PUA values.

File: terminal/src/main/java/org/jline/terminal/KeyParser.java (hunk 4 above)
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/KeyParser.java:857]: Correct — removed the unreachable cases. See the reply above for the full fix: replaced with PUA code mappings (57348–57355) as defensive handling, plus added Home/End.

**Review by gnodet — COMMENTED** [review]
**LGTM** — Re-review after author updates. All three issues from the previous review have been properly addressed:

1. **Dead code in mapKittySpecialKey()** → replaced with correct PUA code mappings (57348-57355) plus Home/End, with clear Javadoc explaining the defensive mapping rationale
2. **Escape binding scope** → now correctly scoped to VIINS and VICMD keymaps only, not applied to emacs
3. **Kitty stack balance** → `setKittyKeyboardMode()` now pops before pushing when already active, keeping the terminal stack at depth 1

Test coverage is thorough: 47 tests covering the complete kitty protocol surface including the newly added PUA navigation key tests and non-BMP character handling.

One minor nit: the Javadoc for `getKeyCode()` says "@return the key code, or 0 if not from a kitty protocol sequence" but legacy character events will return a non-zero keyCode (the character's codepoint). Consider updating the doc to reflect the actual behavior.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: terminal/src/main/java/org/jline/terminal/KeyEvent.java [hunk 6]
```diff
@@ -115,70 +162,102 @@ public enum Modifier {
...
     public KeyEvent(char character, EnumSet<Modifier> modifiers, String rawSequence) {
-        this.type = Type.Character;
-        this.character = character;
-        this.arrow = null;
-        this.special = null;
-        this.functionKey = 0;
-        this.modifiers = modifiers;
-        this.rawSequence = rawSequence;
+        this(Type.Character, character, null, null, 0, modifiers, rawSequence, EventType.Press, character, 0, 0, null);
     }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/KeyEvent.java:292]: Fixed — the legacy character convenience constructor now passes 0 for keyCode instead of the character value, keeping the contract that getKeyCode() returns 0 for non-kitty events.

[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/KeyEvent.java:292]: Fixed — the legacy character convenience constructor now passes 0 for keyCode instead of the character value, keeping the contract that getKeyCode() returns 0 for non-kitty events.

**Review by gnodet — COMMENTED** [review]
**LGTM** — Re-reviewed after force-push. The only change since the last approved review is a 1-line fix in `KeyEvent.java`: the legacy character constructor now passes `0` instead of `character` for the `keyCode` parameter, making `getKeyCode()` behavior match its documented contract ("returns 0 if not from a kitty protocol sequence"). This was the one remaining nit from the prior review and is correctly resolved. No other files changed.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: terminal/src/main/java/org/jline/terminal/impl/KittyKeyboardSupport.java [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+
+    // Navigation keys — PUA codepoints defined by the kitty protocol.
+    // These keys normally use legacy CSI number ~ encoding, but we define the
+    // PUA codes for completeness and to handle terminals that may send them as CSI u.
+    public static final int KEY_INSERT = 57348;
+    public static final int KEY_DELETE = 57349;
+    public static final int KEY_HOME = 57350;
+    public static final int KEY_END = 57351;
+    public static final int KEY_PAGE_UP = 57354;
+    public static final int KEY_PAGE_DOWN = 57355;
```
Comment (gnodet) [inline terminal/src/main/java/org/jline/terminal/impl/KittyKeyboardSupport.java:104]: Per the kitty spec (`key_encoding.py`), `KEY_HOME` (57350) and `KEY_END` (57351) use the PUA codes for LEFT and RIGHT arrow keys. The correct values are HOME=57356 and END=57357. The arrow key codes (57350–57353) are also missing.

This propagates to `mapKittySpecialKey()` in `KeyParser.java` and the test assertions in `KeyParserKittyTest`.

```suggestion
    public static final int KEY_INSERT = 57348;
    public static final int KEY_DELETE = 57349;
    public static final int KEY_LEFT = 57350;
    public static final int KEY_RIGHT = 57351;
    public static final int KEY_UP = 57352;
    public static final int KEY_DOWN = 57353;
    public static final int KEY_PAGE_UP = 57354;
    public static final int KEY_PAGE_DOWN = 57355;
    public static final int KEY_HOME = 57356;
    public static final int KEY_END = 57357;
```

File: terminal/src/main/java/org/jline/terminal/impl/KittyKeyboardSupport.java [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+     * Converts a set of {@link KittyKeyboardMode} values to the protocol bitmask.
+     *
+     * @param modes the enhancement modes
+     * @return the bitmask value used in the protocol escape sequences
+     */
+    @SuppressWarnings("java:S1319") // EnumSet matches JLine API convention (see Attributes)
+    public static int toFlags(EnumSet<KittyKeyboardMode> modes) {
+        int flags = 0;
+        for (KittyKeyboardMode mode : modes) {
+            flags |= (1 << mode.ordinal());
```
Comment (gnodet) [inline terminal/src/main/java/org/jline/terminal/impl/KittyKeyboardSupport.java]: Minor design note: `toFlags()` uses `1 << mode.ordinal()` to derive the wire protocol bitmask. This creates an implicit coupling between the `KittyKeyboardMode` enum declaration order and the spec's flag values (1, 2, 4, 8, 16). Reordering or inserting an enum constant would silently produce wrong values. The `FLAG_*` constants are defined above but not used here.

Not a current bug — the ordinals match today — but worth considering whether to use explicit flag values per constant for robustness.

**Review by gnodet — COMMENTED** [review]
Thorough Kitty Keyboard Protocol implementation with clean integration into JLine's existing patterns. The parser handles all protocol fields correctly, the `bindKittyKeys()` approach of mirroring legacy bindings is the right strategy, and the tests are comprehensive.

Three observations — one medium-severity spec mismatch and two low-severity design notes:

1. **[medium]** `KEY_HOME` and `KEY_END` PUA codepoints are swapped with LEFT/RIGHT arrow codes per the kitty spec (`key_encoding.py`). See inline comment with suggested fix.
2. **[low]** Arrow key PUA codes (LEFT=57350, RIGHT=57351, UP=57352, DOWN=57353) are not mapped anywhere in `buildKittyKeyEvent()` — they fall through to `Type.Unknown`. Defensive mapping would improve completeness.
3. **[low]** `toFlags()` relies on `KittyKeyboardMode.ordinal()` to derive the protocol bitmask — an implicit contract between enum declaration order and flag values.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Comment by crramirez** [comment]
Can be manual tested also on Windows Terminal version 1.25

**Review by gnodet — COMMENTED** [review]
✅ **LGTM** — Well-structured implementation of the Kitty Keyboard Protocol with correct CSI `…u` parsing, proper lifecycle management (push/pop), and comprehensive test coverage (47 new tests). The core functionality is sound.

Observations (informational, non-blocking):

1. **PUA code constants for navigation keys** (`KittyKeyboardSupport.java`): The constants `KEY_INSERT`(57348) through `KEY_PAGE_DOWN`(57355) are not defined in the Kitty Keyboard Protocol spec — PUA assignments start at 57358 (`CAPS_LOCK`). The corresponding `mapKittySpecialKey()` cases in `KeyParser.java` are effectively dead code. If kept for defensive purposes, consider updating the Javadoc to clarify these are JLine-internal mappings rather than spec-defined values.

2. **`KeyEvent.toString()`** (`KeyEvent.java`): `toString()` includes `eventType`, `keyCode`, and `associatedText` but omits `shiftedKeyCode` and `baseLayoutKeyCode`. Including them when non-zero would help debugging.

3. **Test assertion** (`KeyParserKittyTest.java`): `testKeyCodeZero` only asserts `assertNotNull(event)` — verifying the event type would make the test more meaningful.

Overall the API design is clean: push/pop stack model for mode flags, type-safe `KittyKeyboardMode` enum, and backward-compatible `KeyEvent` extension. The `LineReader` integration is minimal and correct.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Labels | _(none)_ | `feature` |
| Milestone | _(none)_ | `5.0.0` |

This is a significant new feature (Kitty Keyboard Protocol) targeting `master`. The `5.0.0` milestone seems appropriate, though `4.4.0` could also work depending on release plans.

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — COMMENTED** [review]
✅ **LGTM** — Re-reviewed after squash + fix commit `fix: correct kitty PUA key codes and decouple toFlags from enum ordinals`.

Both issues from the original review are addressed:

1. **PUA key codes corrected** — Navigation keys (INSERT=57348 through END=57357) now match the kitty spec exactly. Verified against the authoritative kitty source (`key_encoding.py`). The `mapKittyArrowKey()` method correctly maps LEFT=57350, RIGHT=57351, UP=57352, DOWN=57353.

2. **toFlags decoupled from ordinals** — The new `flagFor()` switch maps each `KittyKeyboardMode` to its explicit flag constant (FLAG_DISAMBIGUATE=1 through FLAG_REPORT_TEXT=16), eliminating the fragile `1 << ordinal()` pattern.

**Other improvements noted:**
- Legacy character constructor now passes `0` for keyCode, aligning `getKeyCode()` with its documented contract
- `setKittyKeyboardMode()` now pops before pushing when already active, keeping the terminal stack balanced

**Minor observation (non-blocking):** F1–F12 PUA codepoints (57364–57375) aren't handled by `mapKittyFunctionKey` (only F13–F35). If a terminal sends F1 as `CSI 57364u`, it would classify as Unknown. Very low impact since the spec says F1–F12 should use legacy CSI encoding.

Test coverage is thorough with 47+ tests. SonarCloud QG passed.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|-----------|
| Labels | _(none)_ | `enhancement` |
| Milestone | _(none)_ | `4.4.0` |

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._
_Claude Code on behalf of Guillaume Nodet_

File: .mvx/demo.sh [hunk 9]
```diff
@@ -39,7 +39,7 @@ fi
 TARGETDIR="demo/target"
 cp="${TARGETDIR}/classes"
 logconf="demo/etc/logging.properties"
-JVM_OPTS=""
+JVM_OPTS="--enable-native-access=ALL-UNNAMED"
```
[the remark this answers is not in the document]
Comment (gnodet) [reply .mvx/demo.sh]: Do we still support JDK < 11 ?

**Review by gnodet — COMMENTED** [review]
✅ **LGTM** — Well-structured implementation of the Kitty Keyboard Protocol with 75 tests, clean lifecycle management (push/pop stack model), and correct integration into JLine's existing patterns. All 14 CI checks pass. The design decision to only enable flag 1 (Disambiguate) by default is pragmatic.

**Non-blocking observations:**

1. **`parseTextCodepoints` codepoint validation** (medium) — `KeyParser.java` line 670: A malformed terminal response with a codepoint > `0x10FFFF` will throw `IllegalArgumentException` from `StringBuilder.appendCodePoint()`. Adding `&& Character.isValidCodePoint(cp)` to the existing `if (cp > 0)` guard would make this more robust.

2. **`parseModifierCode` extended modifiers** (low) — Line 435: For legacy CSI-format arrow/F-keys with kitty mode active, extended modifiers (Super/Hyper/Meta/CapsLock/NumLock) would be silently dropped since only 3 bits are decoded. Consider reusing `parseKittyModifiers` there.

3. **F1-F12 PUA codepoints** (low) — Line 924: `mapKittyFunctionKey` only handles F13-F35 PUA codes (57376-57398). If a terminal sends F1 as `CSI 57364u`, it would classify as Unknown. Low real-world impact since the spec says terminals should use legacy CSI encoding for F1-F12.

4. **`testKeyCodeZero`** (low) — Only asserts `assertNotNull(event)` without verifying the event type or fields.

5. **Missing keypad/media codes** (low) — KP_BEGIN (57427) and several media key codes (57431-57440) from the spec are not mapped.

None of these are blocking — the PR has been through extensive review iterations and all prior issues have been addressed.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Category | _(unlabeled)_ | `enhancement` |
| Labels | _(none)_ | + `enhancement` |
| Milestone | _(none)_ | `4.4.0` |

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
